File: src/solutions/python/SumofTwoIntegers.py

'''
 # Problem
 	- `Link`: https://leetcode.com/problems/sum-of-two-integers/
 # Solution
 	- `Author`: Kyungtaek Lim (Jonas)
 	- `Date`: Mar 26, 2025
 	- `Answer`: getSum
 # Reference
	- https://github.com/kyungtaek-jonas-lim/jonascode/blob/main/doc/reference/BitwiseOperators.md
'''
class Solution:

    '''
    Sum
    - O(1)
    - https://github.com/kyungtaek-jonas-lim/jonascode/blob/main/doc/explanation/GetSum.md
    '''
    def getSum(self, a: int, b: int) -> int:
        while b != 0:
            carry = a & b
            a = a ^ b
            b = carry << 1
        return a
    
    '''
    Minus
    '''
    def get_difference(self, a: int, b: int) -> int:
        MAX = 0xFFFFFFFF
        MASK = 0x7FFFFFFF  # For handling negative numbers

        # Continue until there is no borrow left
        while b != 0:
            # Identify borrow bits: where a has 0 and b has 1
            borrow = (~a) & b

            # XOR computes difference without borrow
            a = a ^ b

            # Shift borrow left by 1 to apply it to the next bit position
            b = borrow << 1

            # Simulate 32-bit integer
            a &= MAX
            b &= MAX

        # If a is negative in 32-bit, convert it to Python negative
        return a if a <= MASK else ~(a ^ MAX)


    '''
    Production
    '''
    def get_product(self, a: int, b: int) -> int:
        result = 0

        # Repeat while b is not zero
        while b != 0:
            # If the least significant bit of b is 1, add 'a' to result
            if b & 1:
                result += a

            # Left shift 'a' (equivalent to a * 2), right shift 'b' (equivalent to b // 2)
            a <<= 1
            b >>= 1

        return result


    '''
    Division
    '''
    def get_quotient(self, a: int, b: int) -> int:
        if b == 0:
            raise ZeroDivisionError("Division by zero")

        result = 0
        # Determine the sign of the result
        sign = -1 if (a < 0) ^ (b < 0) else 1

        a = abs(a)
        b = abs(b)

        # Subtract shifted divisor from dividend until dividend is smaller
        while a >= b:
            temp = b
            multiple = 1

            # Shift divisor left until it's just below or equal to 'a'
            while a >= (temp << 1):
                temp <<= 1
                multiple <<= 1

            a -= temp
            result += multiple

        return sign * result


if __name__ == '__main__':
    sol = Solution()
    print(sol.getSum(1, 2)) # 3
    print(sol.getSum(2, 3)) # 5

    print("---")
    print(sol.get_difference(5, 3))     # 2
    print(sol.get_difference(1, 3))     # -2
    print(sol.get_difference(100, 200)) # -100
    print(sol.get_difference(7, 0))     # 7
    print(sol.get_difference(0, 7))     # -7
    print(sol.get_difference(-3, -1))   # -2


    print("---")
    print(sol.get_product(1, 2)) # 2
    print(sol.get_product(2, 3)) # 6

    print("---")
    print(sol.get_quotient(1, 2)) # 0
    print(sol.get_quotient(2, 3)) # 0

File: src/solutions/java/SumofTwoIntegers.java
package solutions.java;

import java.util.Scanner;

/*
 # Problem
 	- `Link`: https://leetcode.com/problems/sum-of-two-integers/
 # Solution
 	- `Author`: Kyungtaek Lim (Jonas)
 	- `Date`: Jan 4, 2025
 	- `Answer`: getSum / getSumAdvanced
 # Reference
	- https://github.com/kyungtaek-jonas-lim/jonascode/blob/main/doc/reference/BitwiseOperators.md
 */

public class GetSum {
	public static void main(String[] args) {
        
        Scanner scanner = new Scanner(System.in);
        
        // Input the first and second numbers 
        System.out.print("Enter the first integer: ");
        int first = scanner.nextInt();
        System.out.print("Enter the second integer: ");
        int second = scanner.nextInt();
		
//        final int sum = getSum(first, second);
        final int sum = getSumAdvanced(first, second);        
		System.out.println("[RESULT] The sum is `" + sum + "`");
		
		scanner.close();
	}

	/*
	 * Option #1 
	 * Common way
	 * O(n)
	 */
    public static int getSum(int a, int b) {
    	int num = 0;
        for (int i = 0; i < a; i++) {
        	num++;
        }
        
        for (int i = 0; i < b; i++) {
        	num++;
        }
        return num;
    }
    
    /*
     * Option #2
     * Advanced way
     * O(1)
     * https://github.com/kyungtaek-jonas-lim/jonascode/blob/main/doc/explanation/GetSum.md
     */
    public static int getSumAdvanced(int a, int b) {
    	while (b != 0) {
    		// 1. Calculate carry (common bits that are set in both a and b)
    		int carry = a & b;
    		// 2. Perform XOR operation to add without carry 
    		a = a ^ b;
    		// Shift carry left by 1 so that it adds to the next higher bit
    		b = carry << 1;
    	}
    	return a;
    }

    // ===================================================
    // Minus: Extra Function
    // Instead of "*"
    public static int getDifference(int a, int b) {
        // Continue until there is no borrow left
        while (b != 0) {
            // Calculate the borrow by using NOT on a and AND with b
            // This identifies which bits in a need to be borrowed
            int borrow = (~a) & b;  // Borrow bits where a is 0 and b is 1

            // XOR is used to compute the difference without borrow
            // It works by subtracting where there's no need to borrow
            a = a ^ b;

            // Shift the borrow left by 1 to add it to the next higher bit
            b = borrow << 1;  // Moving the borrow to the next bit
        }
        return a;  // Return the final result after handling all borrows
    }

    // ===================================================
    // Production: Extra Function
    // Instead of "*"
    public int getProduct(int a, int b) {
        int result = 0;

        // Perform multiplication using bitwise shift and add
        while (b != 0) {
            // If the least significant bit of b is 1, add a to the result
            if ((b & 1) == 1) {
                result = result + a;
            }

            // Shift a to the left (multiply by 2) and b to the right (divide by 2)
            a = a << 1;  // a * 2
            b = b >> 1;  // b / 2
        }
        return result;
    }
    
    // ===================================================
    // Division: Extra Function
    // Instead of "/" and "%"
    public int getQuotient(int a, int b) {
        // Handle edge case when b is 0
        if (b == 0) {
            throw new ArithmeticException("Division by zero");
        }

        int result = 0;
        int sign = (a < 0) ^ (b < 0) ? -1 : 1;  // Determine the sign of the result

        a = Math.abs(a);  // Work with absolute values of a and b
        b = Math.abs(b);

        // Perform division using bitwise shift
        while (a >= b) {
            int temp = b, multiple = 1;
            
            // Left shift the divisor until it's larger than the dividend
            while (a >= (temp << 1)) {
                temp <<= 1;
                multiple <<= 1;  // Track how many times we shifted
            }

            // Subtract the shifted divisor from the dividend and add the multiple
            a -= temp;
            result += multiple;
        }

        return result * sign;  // Apply the sign to the result
    }

}
